File: src/07.greedy/01.arrayGreedy/22.517.go
package array_greedy

func FindMinMoves(machines []int) int {
	n := len(machines)
	sum := 0
	for _, v := range machines {
		sum += v
	}

	if sum%n != 0 {
		return -1
	}

	avg := sum / n
	ls := 0
	rs := sum
	ans := 0

	for i, v := range machines {
		rs -= v
		a := avg*i - ls
		if a < 0 {
			a = 0
		}

		b := (n-i-1)*avg - rs
		if b < 0 {
			b = 0
		}

		if ans < a+b {
			ans = a + b
		}
		ls += v
	}

	return ans
}

File: src/07.greedy/01.arrayGreedy/22.517.ts
export function findMinMoves(machines: number[]): number {
    // 无解情况
    // 衣服数量不能整除洗衣机数量 返回 -1

    const n = machines.length;
    let sum = 0;
    for (let v of machines) {
        sum += v;
    }

    if (sum % n !== 0) {
        return -1;
    }

    // 每个洗衣机中的衣服数量
    let t = sum / n;

    // 左侧衣服数量
    let ls = 0;

    // 右侧衣服数量
    let rs = sum;
    let ans = 0;

    for (let i = 0; i < n; i++) {
        rs -= machines[i];

        const a = Math.max(t * i - ls, 0);
        const b = Math.max((n - i - 1) * t - rs, 0);
        ans = Math.max(ans, a + b);
        ls += machines[i];
    }

    return ans;
}
